### PR TITLE
WL-286 Add support for multitenant OAuth

### PR DESCRIPTION
**Pull request by douglashall**
- Pull URL references from SiteConfiguration instead of settings.
- Upgrade edx-auth-backends to 0.2.0 https://github.com/edx/auth-backends/pull/9

Sandbox URLs:
https://ecommerce-multitenant.sandbox.edx.org/
https://ecommerce-hms-multitenant.sandbox.edx.org/
https://ecommerce-mit-multitenant.sandbox.edx.org/

### CODE REVIEW

**Comment by douglashall** [comment]
@clintonb This is ready for a final review. The tests will not run on Travis until be get edx-auth-backends==0.2.0 pushed up to pypi. They do run successfully locally with auth-backends:douglashall/WL-286/support_multi_tenancy source installed.

File: ecommerce/social_auth/strategies.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+class CurrentSiteDjangoStrategy(DjangoStrategy):
+    """
+    Python Social Auth strategy which accounts for the current
+    Site when enabling third party authentication.
+    """
+
+    def get_setting(self, name):
+        # First check the current SiteConfiguration for the setting
+        setting = self.request.site.siteconfiguration.social_auth_settings.get(name)
+        if not setting:
```
Comment (clintonb) [inline ecommerce/social_auth/strategies.py:15]: This doesn't seem right. We should rely solely on site configuration. Having two sources of data has the potential to lead to issues with debugging. Also, it will be nice to get rid of these unnecessary settings.

Comment (douglashall) [inline ecommerce/social_auth/strategies.py:15]: We are going to need this to roll this change out to production. We can go back and fully rely on SiteConfiguration later on.

Comment (clintonb) [inline ecommerce/social_auth/strategies.py:15]: We can deploy to a single cluster outside of the ELB. Once the settings are in place, we can cut over.

Comment (douglashall) [inline ecommerce/social_auth/strategies.py:15]: If you are signing up to manage that deployment, then I can come onboard. Otherwise, I would prefer to stick with what we have here for now.

Comment (clintonb) [inline ecommerce/social_auth/strategies.py:15]: This is the cleanest deployment strategy. Say we keep this as written. How do you know your changes from `SiteConfiguration` are being used instead of the standard Django settings? As is, there is no assurance the new path/settings are being used. The full switchover will require some manual intervention, but we will know it works.

Comment (clintonb) [inline ecommerce/social_auth/strategies.py:15]: Hmm...even if we go with the full cutover, this will need to be deployed in two phases: (1) add column to configuration, (2) switch to using new strategy. If we do the full cutover, there will be no way to log into the server. However, we could define a password for someone's staff account and bypass OIDC login altogether. Either way, you have options. My preference is that we make a full cutover rather than take a hybrid approach.

Comment (douglashall) [inline ecommerce/social_auth/strategies.py:15]: It will work well as written. Not sure what you mean by "no assurance the new path/settings are being used". The new sites that we add will not work at all if we are not pulling settings from SiteConfiguration. The existing site will work because it will be using this fallback to grab settings from django.

Comment (douglashall) [inline ecommerce/social_auth/strategies.py:15]: I am happy to create a follow up Jira task to remove the fallback and pull it into our next sprint if the concern is that we will never remove the fallback.

Comment (clintonb) [inline ecommerce/social_auth/strategies.py:15]: Cool.

Comment (douglashall) [inline ecommerce/social_auth/strategies.py:15]: https://openedx.atlassian.net/browse/WL-333

**Comment by douglashall** [comment]
@mattdrayer Could you give this a second review?

File: ecommerce/core/admin.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 
 class SiteConfigurationAdmin(admin.ModelAdmin):
-    list_display = ('site', 'partner', 'lms_url_root', 'theme_scss_path', 'payment_processors')
+    list_display = ('site', 'partner', 'lms_url_root', 'theme_scss_path', 'payment_processors', 'social_auth_settings')
```
Comment (clintonb) [inline ecommerce/core/admin.py]: These settings don't need to be in the list display.

File: ecommerce/core/context_processors.py [hunk 3]
```diff
@@ -1,8 +1,11 @@
 from django.conf import settings
 
+from ecommerce.core import get_lms_dashboard_url
+
 
 def core(_request):
     return {
+        'lms_dashboard_url': get_lms_dashboard_url(),
+        'platform_name': settings.PLATFORM_NAME,
```
Comment (clintonb) [inline ecommerce/core/context_processors.py:9]: This will eventually need to be changed to either the site or partner name.

File: ecommerce/core/models.py [hunk 4]
```diff
@@ -34,6 +34,13 @@ class SiteConfiguration(models.Model):
         null=False,
         blank=False
     )
+    social_auth_settings = JSONField(
+        verbose_name=_('Python Social Auth settings'),
+        help_text=_("JSON string containing Python Social Auth backend settings."),
+        max_length=255,
```
Comment (clintonb) [inline ecommerce/core/models.py]: There shouldn't be a maximum length here.

File: ecommerce/core/models.py [hunk 5]
```diff
@@ -34,6 +34,13 @@ class SiteConfiguration(models.Model):
...
+        null=False,
```
Comment (clintonb) [inline ecommerce/core/models.py:40]: Unless you specify a default, this won't work.

File: ecommerce/core/models.py [hunk 6]
```diff
@@ -34,6 +34,13 @@ class SiteConfiguration(models.Model):
         null=False,
         blank=False
     )
+    social_auth_settings = JSONField(
```
Comment (clintonb) [inline ecommerce/core/models.py]: We shouldn't tie our field names to third-party libraries. What if we decide to replace PSA? `oauth_settings` is more appropriate. The `verbose_name` and `help_text` values should be modified as well.

File: ecommerce/core/tests/test_views.py [hunk 7]
```diff
@@ -33,6 +34,8 @@ def test_all_services_available(self, mock_lms_request):
 
         self._assert_health(status.HTTP_200_OK, Status.OK, Status.OK, Status.OK)
 
+    @mock.patch('ecommerce.core.views.get_lms_heartbeat_url', mock.Mock(return_value=''))
```
Comment (clintonb) [inline ecommerce/core/tests/test_views.py:37]: Why are these mocks only necessary for the database test? If they must be mocked, more realistic values should be used.

File: ecommerce/coupons/tests/test_views.py [hunk 8]
```diff
@@ -129,7 +129,6 @@ def test_used_voucher(self):
         self.assertEqual(msg, _('This coupon has already been used'))
 
 
-@httpretty.activate
```
Comment (clintonb) [inline ecommerce/coupons/tests/test_views.py:143]: Why was this removed? `httpretty` is used by almost every test in this class.

File: ecommerce/coupons/tests/test_views.py [hunk 9]
```diff
@@ -11,10 +11,10 @@
 from oscar.test.utils import RequestFactory
```
Comment (clintonb) [inline ecommerce/coupons/tests/test_views.py:11]: What's going on in this file? Aside form changing `get_lms_url`, why are any of the other changes necessary?

File: ecommerce/extensions/api/data.py [hunk 10]
```diff
@@ -59,8 +59,9 @@ def get_lms_footer():
         str: HTML representation of the footer.
     """
     try:
+        lms_url = get_lms_url('api/branding/v1/footer')
```
Comment (clintonb) [inline ecommerce/extensions/api/data.py]: `footer_api_url`

File: ecommerce/extensions/api/serializers.py [hunk 11]
```diff
@@ -373,8 +373,7 @@ def get_benefit(self, obj):
         return (obj.offers.first().benefit.type, obj.offers.first().benefit.value)
 
     def get_redeem_url(self, obj):
-        domain = settings.ECOMMERCE_URL_ROOT
-        return "{}/coupons/offer/?code={}".format(domain, obj.code)
+        return get_ecommerce_url("/coupons/offer/?code={}".format(obj.code))
```
Comment (clintonb) [inline ecommerce/extensions/api/serializers.py]: Append the query string after you get the URL. `get_ecommerce_url` expects to receive a path, so that's all it should receive.

``` python
def get_redeem_url(self, obj):
  url = get_ecommerce_url('/coupons/offer/')
  return '{url}?code={code}'.format(url=url, code=obj.code)
```

File: ecommerce/extensions/dashboard/users/views.py [hunk 12]
```diff
@@ -1,11 +1,14 @@
 import logging
+import requests
```
Comment (clintonb) [inline ecommerce/extensions/dashboard/users/views.py]: Put these back where they were. Neither is a default Python library.

File: ecommerce/extensions/voucher/utils.py [hunk 13]
```diff
@@ -77,7 +79,7 @@ def generate_coupon_report(coupon_vouchers):
                 discount = _("{percentage} %").format(percentage=benefit_value)
             else:
                 discount = currency(benefit_value)
-            URL = '{}/coupons/redeem/?code={}'.format(settings.ECOMMERCE_URL_ROOT, voucher.code)
+            URL = get_ecommerce_url('/coupons/redeem/?code={}'.format(voucher.code))
```
Comment (clintonb) [inline ecommerce/extensions/voucher/utils.py]: Same as above. Pass only the path to `get_ecommerce_url`.

File: ecommerce/settings/base.py [hunk 14]
```diff
@@ -209,12 +211,6 @@
 # Absolute URL used to construct LMS URLs.
 LMS_URL_ROOT = None
```
Comment (clintonb) [inline ecommerce/settings/base.py]: Is this still needed?

File: ecommerce/settings/test.py [hunk 15]
```diff
@@ -45,15 +47,9 @@ def __getitem__(self, item):
 
 
 # URL CONFIGURATION
-ECOMMERCE_URL_ROOT = 'http://localhost:8002'
-
-LMS_URL_ROOT = 'http://127.0.0.1:8000'
+ECOMMERCE_URL_ROOT = 'http://testserver.fake'
```
Comment (clintonb) [inline ecommerce/settings/test.py]: This should not be the same as `LMS_URL_ROOT`. Make them different to avoid confusion when debugging test failures.

File: ecommerce/settings/test.py [hunk 16]
```diff
@@ -70,6 +66,9 @@ def __getitem__(self, item):
 PASSWORD_HASHERS = (
     'django.contrib.auth.hashers.MD5PasswordHasher',
 )
+
+SOCIAL_AUTH_EDX_OIDC_MULTISITE_KEY = 'fake-key'
```
Comment (clintonb) [inline ecommerce/settings/test.py]: These aren't needed.

File: ecommerce/social_auth/tests/test_strategies.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+from ecommerce.tests.testcases import TestCase
+
+
+class CurrentSiteDjangoStrategyTests(TestCase):
+    """Tests of the CurrentSiteDjangoStrategy."""
+
+    def setUp(self):
+        super(CurrentSiteDjangoStrategyTests, self).setUp()
+        self.site.siteconfiguration.social_auth_settings = {
+            'SOCIAL_AUTH_EDX_OIDC_MULTISITE_KEY': 'test-key'
```
Comment (clintonb) [inline ecommerce/social_auth/tests/test_strategies.py]: MULTISITE is no longer valid.

File: ecommerce/tests/mixins.py [hunk 18]
```diff
@@ -247,6 +249,7 @@ def setUp(self):
 
         Site.objects.get_current().delete()
         site_configuration = SiteConfigurationFactory(
+            lms_url_root="http://%s" % domain,
```
Comment (clintonb) [inline ecommerce/tests/mixins.py]: Move this to the factory. Also, use 'https' and `string.format()`. See the decorator pattern for lazy attributes at https://factoryboy.readthedocs.org/en/latest/introduction.html#lazyattribute, and pull the value from `obj.site.domain`.

File: ecommerce/core/context_processors.py (hunk 3 above)
Comment (douglashall) [inline ecommerce/core/context_processors.py:9]: Yes, let's make that change separately.

File: ecommerce/core/tests/test_views.py (hunk 7 above)
Comment (douglashall) [inline ecommerce/core/tests/test_views.py:37]: Because by mocking the database cursor, we cause the Django CurrentSiteMiddleware to fail when it tries to look up the current site. Not sure what you mean by more realistic value?

Comment (clintonb) [inline ecommerce/core/tests/test_views.py:37]: "more realistic value" -> an actual URL

Comment (douglashall) [inline ecommerce/core/tests/test_views.py:37]: The URL returned by the mock is inconsequential considering the purpose of this test.

File: ecommerce/coupons/tests/test_views.py (hunk 9 above)
Comment (douglashall) [inline ecommerce/coupons/tests/test_views.py:11]: For some reason we were requiring an LMS instance to be up and running in order for most of these tests to run. I have fixed this by mocking the calls to the LMS footer endpoint.

File: ecommerce/coupons/tests/test_views.py (hunk 8 above)
Comment (douglashall) [inline ecommerce/coupons/tests/test_views.py:143]: Decorating the individual test methods seems more clear when every test does not have a request being mocked.

File: ecommerce/coupons/tests/test_views.py (hunk 9 above)
Comment (clintonb) [inline ecommerce/coupons/tests/test_views.py:11]: Something's wrong here. These tests did not require an LMS instance. If they had they never would have been merged, because Travis doesn't run LMS.

Comment (clintonb) [inline ecommerce/coupons/tests/test_views.py:11]: Taking a closer look, `CouponOfferViewTests.setUp()` mocks the call to the LMS Footer API.

Comment (douglashall) [inline ecommerce/coupons/tests/test_views.py:11]: Right. I will revert these changes.

File: ecommerce/settings/base.py (hunk 14 above)
Comment (douglashall) [inline ecommerce/settings/base.py]: Yes. There are tests that reference this setting.

Comment (clintonb) [inline ecommerce/settings/base.py]: Which tests? Why not update them and remove this setting, especially since there's potential for brand leakage?

Comment (douglashall) [inline ecommerce/settings/base.py]: I have updated the tests to use `ecommerce.core.get_lms_url()` instead of the setting. There is still a management command that wants the setting. Any objections to making the LMS URL an option on the command?

https://github.com/edx/ecommerce/blob/douglashall/WL-286/support_multitenant_oauth/ecommerce/extensions/catalogue/management/commands/migrate_course.py#L42

Comment (clintonb) [inline ecommerce/settings/base.py]: `migrate_course`? It takes a `Partner` identifier. Use that to get the `SiteConfiguration` and the correct LMS URL.

Comment (douglashall) [inline ecommerce/settings/base.py]: `Partner` can have more than one `SiteConfiguration` though, no?

Comment (clintonb) [inline ecommerce/settings/base.py]: Yeah...replace `Partner` with `Site`.

**Comment by douglashall** [comment]
@clintonb This is ready for another look. I have removed all references to URLs in Django settings.

File: ecommerce/core/__init__.py [hunk 19]
```diff
@@ -1 +1,47 @@
...
+
 default_app_config = 'ecommerce.core.config.CoreAppConfig'  # pragma: no cover
+
+
+def get_ecommerce_url(path=''):
+    request = get_current_request()
+    if request:
+        ecommerce_url_root = "{}://{}".format(request.scheme, request.site.domain)
+        return urljoin(ecommerce_url_root, path)
+    return None
```
Comment (clintonb) [inline ecommerce/core/__init__.py]: An exception (e.g. `MissingRequestError`) should be raised if there is no `request` value. Otherwise, you run the risk of consuming code not checking the return value, and displaying an invalid URL. Also, add a docstring.

File: ecommerce/core/__init__.py [hunk 20]
```diff
@@ -1 +1,47 @@
...
+    if request:
+        return urljoin(request.site.siteconfiguration.lms_url_root, path)
+    return None
+
+
+def get_oauth2_provider_url():
+    return get_lms_url('/oauth2')
+
+
+def rewrite_lms_url_for_current_site(url):
```
Comment (clintonb) [inline ecommerce/core/__init__.py]: Update the payment providers settings to accept a receipt/error page path instead of a URL.

File: ecommerce/core/tests/test_init.py [hunk 21]
```diff
@@ -0,0 +1,10 @@
+import mock
```
Comment (clintonb) [inline ecommerce/core/tests/test_init.py]: Consider moving the code being tested out of `__init__.py` and into `url_utils.py`. Rename this file to `test_url_utils.py`.

File: ecommerce/core/tests/test_init.py [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+
+from ecommerce.core import get_lms_url
+from ecommerce.tests.testcases import TestCase
+
+
+class UrlUtilityTests(TestCase):
+    @mock.patch('ecommerce.core.get_current_request', mock.Mock(return_value=None))
+    def test_get_lms_url_with_no_current_request(self):
```
Comment (clintonb) [inline ecommerce/core/tests/test_init.py]: What if there _is_ a request? The positive case needs to be tested.

Comment (douglashall) [inline ecommerce/core/tests/test_init.py]: The positive case is already covered by numerous other tests.

File: ecommerce/courses/publishers.py [hunk 23]
```diff
@@ -74,8 +74,9 @@ def publish(self, course, access_token=None):
             course_id=course_id
         )
 
-        if not settings.COMMERCE_API_URL:
-            logger.error('COMMERCE_API_URL is not set. Commerce data will not be published!')
+        commerce_api_url = get_lms_commerce_api_url()
+        if not commerce_api_url:
+            logger.error('Could not find the commerce api url. Commerce data will not be published!')
```
Comment (clintonb) [inline ecommerce/courses/publishers.py]: Commerce API URL si not set. Commerce data will not be published.

File: ecommerce/core/tests/test_init.py (hunk 21 above)
Comment (douglashall) [inline ecommerce/core/tests/test_init.py]: I would like to leave these in `__init__.py`

File: ecommerce/core/__init__.py (hunk 20 above)
Comment (douglashall) [inline ecommerce/core/__init__.py]: The idea here was to make for a smooth rollout. Do we store the payment processor settings in https://github.com/edx/configuration somewhere? Can you point me to the right place?

File: ecommerce/extensions/catalogue/management/commands/migrate_course.py [hunk 24]
```diff
@@ -171,30 +165,30 @@ class Command(BaseCommand):
...
                     help='Save the migrated data to the database. If this is not set, '
                          'migrated data will NOT be saved to the database.'),
-        make_option('--partner',
+        make_option('--site',
                     action='store',
-                    dest='partner_short_code',
+                    dest='site_domain',
                     default=None,
-                    help='Short code for the partner providing the course.'),
+                    help='Domain for the site providing the course.'),
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/management/commands/migrate_course.py]: ecommerce site

File: ecommerce/settings/production.py [hunk 25]
```diff
@@ -11,6 +11,10 @@
 from ecommerce.settings.logger import get_logger_config
 
 
+# See: https://docs.djangoproject.com/en/dev/ref/settings/#site-id
+# This needs to be set to None in order to support multitenancy
+SITE_ID = None
```
Comment (clintonb) [inline ecommerce/settings/production.py:16]: The value is not defined by default. Is this really necessary?

Comment (douglashall) [inline ecommerce/settings/production.py:16]: Correct. I'm erring on the side of over documentation here.

File: ecommerce/core/__init__.py (hunk 20 above)
Comment (clintonb) [inline ecommerce/core/__init__.py]: https://github.com/edx/configuration/blob/master/playbooks/roles/ecommerce/defaults/main.yml#L62-L80

File: ecommerce/core/tests/test_init.py (hunk 21 above)
Comment (clintonb) [inline ecommerce/core/tests/test_init.py]: Why?

Comment (douglashall) [inline ecommerce/core/tests/test_init.py]: Makes for shorter imports. Is there a reason we would want to move them?

File: ecommerce/core/__init__.py (hunk 19 above)
Comment (douglashall) [inline ecommerce/core/__init__.py]: Do we have a module that houses custom errors somewhere? I can't seem to find one.

File: ecommerce/core/tests/test_init.py (hunk 21 above)
Comment (clintonb) [inline ecommerce/core/tests/test_init.py]: Organization. All of the functions are URL utilities. Grouping them into a package seems logical. Shorter imports don't really matter.

File: ecommerce/core/__init__.py (hunk 19 above)
Comment (clintonb) [inline ecommerce/core/__init__.py]: Custom exceptions generally go into an exceptions package. In this case, you'll want to create `ecommerce/core/exceptions.py`.

File: ecommerce/core/tests/test_init.py (hunk 21 above)
Comment (douglashall) [inline ecommerce/core/tests/test_init.py]: Sounds good. This is something we would want to raise earlier in the review process.

**Comment by douglashall** [comment]
@clintonb All comments addressed. Thanks again for all your help.

File: ecommerce/core/url_utils.py [hunk 26]
```diff
@@ -0,0 +1,46 @@
+from urlparse import urljoin
+
+from threadlocals.threadlocals import get_current_request
+
+from ecommerce.core.exceptions import MissingRequestError
+
+
+def get_ecommerce_url(path=''):
+    """
+    Returns path joined with the appropriate ecommerce URL root for the current site
```
Comment (clintonb) [inline ecommerce/core/url_utils.py:10]: Document the exception being raised.

**Comment by clintonb** [comment]
:+1:
